---
title: "Parentage_apparent"
author: "Sam Gurr"
date: "2024-04-25"
output: html_document
---

## Set the working directory

```{r setup, include=FALSE}
knitr::opts_knit$set(root.dir = "C:/Users/samuel.gurr/Documents/Github_repositories/EAD-ASEB-Airradians_multigen_OA") # sets the working directo
knitr::opts_knit$set(root.dir = "C:/Users/samjg/Documents/Github_repositories/Airradians_multigen_OA") # sets the working directo

```

## Load Libraries 
```{r, load libraries}
library(outliers)
library(dplyr)
library(stringi)
library(tibble)
library(stringr)
library(BEDASSLE)

```

```{r call data, include=FALSE}
path = "/HPC_analysis/output/lcWGS/angsd/"
# F0 Broodstock
F0_raw_geno <- read.table('HPC_analysis/output/lcWGS/angsd/F0_Broodstock/F0Broodstock_doMaf1_minMaf0.05_majorminor4_minind25_minD5x125_maxD20x500minDind5_maxDind150_minq30_minmapQ.geno.gz')  
F0_IDs       <- read.table('HPC_analysis/output/lcWGS/angsd/F0_Broodstock/ID') %>% 
                    dplyr::mutate(V1 = sub('.*adapter_trim.', '', V1))
F0_parentage <- read.csv('HPC_analysis/output/lcWGS/angsd/F0_Broodstock/parentage.csv') %>% 
                    dplyr::mutate(Sample.ID = sub('.*adapter_trim.', '', Sample.ID))
getwd()
# F1 Juveniles pH 8 
F1J_pH8_raw_geno <- read.table('HPC_analysis/output/lcWGS/angsd/F1_Juveniles/F1_Juveniles_8/F1Juveniles_pH8_doMaf1_minMaf0.05_majorminor4_minind35_minD5x175_maxD20x700minDind5_maxDind150_minq30_minmapQ.geno.gz')  
F1J_pH8_IDs      <- read.table('HPC_analysis/output/lcWGS/angsd/F1_Juveniles/F1_Juveniles_8/IDs.txt') %>% 
                    dplyr::mutate(V1 = str_match(V1, "trim.\\s*(.*?)\\s*_query")[,2])

# F1 Juveniles pH 75
F1J_pH75_raw_geno <- read.table('HPC_analysis/output/lcWGS/angsd/F1_Juveniles/F1_Juveniles_75/F1Juveniles_pH75_doMaf1_minMaf0.05_majorminor4_minind38_minD5x190_maxD20x760minDind5_maxDind150_minq30_minmapQ.geno.gz')  
F1J_pH75_IDs      <- read.table('HPC_analysis/output/lcWGS/angsd/F1_Juveniles/F1_Juveniles_75/IDs.txt') %>% 
                    dplyr::mutate(V1 = str_match(V1, "trim.\\s*(.*?)\\s*_query")[,2])



# create key for downastream merge to master file 


key_F1J_pH75 <- rbind(F0_parentage,  (F1J_pH75_IDs %>% dplyr::mutate(key = 'Off') %>% dplyr::rename(Sample.ID = V1)) )
key_F1J_pH8  <- rbind(F0_parentage,  (F1J_pH8_IDs %>% dplyr::mutate(key = 'Off') %>% dplyr::rename(Sample.ID = V1)) )


```


```{r doGene convert genotype}

converter = function (InputFile) {

  OutputFile  <- InputFile # keep your input instead of overwritting
  
  for (x in 1:nrow(OutputFile)) {
    
    MajorAllele <- OutputFile[x, 3]
    MinorAllele <- OutputFile[x, 4]
    
    
    for (y in 5:ncol(OutputFile)) {
      
      if (OutputFile[x,y] == paste0(MajorAllele,MajorAllele)) { # homozygous major
        
        OutputFile[x,y] = 0 # homozygous major
        
      } else if (OutputFile[x,y] == paste0(MajorAllele,MinorAllele) |
                 OutputFile[x,y] == paste0(MinorAllele,MajorAllele)) { # heterozygous minor 
        
        OutputFile[x,y] = 1 # heterozygous minor 
        
      } else if (OutputFile[x,y] == paste0(MinorAllele,MinorAllele)) { # homozygous minor 
       
         OutputFile[x,y] = 2 # homozygous minor 
      
      } else { # NA, -1 via doGeno in angsd
        
         OutputFile[x,y] = -1
         
      }
        OutputFile[x,y] = as.numeric(OutputFile[x,y])
    } # y cols each individual starting at column 5 via angsd -doGeno = 4 output
    
  } # x rows each SNP chromosome and position
  
  Output <- list(OutputFile)
  return (Output)
  
} # close function

```



```{r run converter Juveniles Mod pCO2}
converter(F1J_pH75_raw_geno)

# View(OutputFile)


# convert all genotypes to numeric
cols <- (colnames(OutputFile[,5:ncol(OutputFile)]))
OutputFile[cols] <- sapply(OutputFile[cols], as.numeric)


OutputFile_formatted <- OutputFile %>% 
                      dplyr::mutate(V1 = paste(V1,V2, sep ='_')) %>% 
                      dplyr::select(!c(V2, V3, V4)) %>% 
                      dplyr::rename(Gene.ID = V1)

# input the IDs as columns 2 through the end of the format_geno
colnames(OutputFile_formatted)[2:ncol(OutputFile_formatted)] = F1J_pH75_IDs$V1
# as matrix 
OutputFile_formatted.asmatrix <- OutputFile_formatted
rownames(OutputFile_formatted.asmatrix)<-OutputFile_formatted.asmatrix$Gene.ID # make Gene.Id into rownames
OutputFile_formatted.asmatrix$Gene.ID<-NULL # omit Gene.ID - 
# tranpose and format rownames back to the first column as Sample.ID 
OutputFile_transposed <- as.data.frame(t(OutputFile_formatted.asmatrix)) %>% 
                        tibble::rownames_to_column("Sample.ID")

nnorm <- function(x) x/sum(x)
getSFS<-function(x) table(factor(rowSums(x[,-c(1:4)]),levels=1:20))

table(factor(rowSums(OutputFile[,-c(1:4)]),levels=1:20))
x <- (rowSums(OutputFile[,-c(1:4)]))
y <- paste0(OutputFile[,1], OutputFile[,2])
data.frame(ncol=length(y))

m <- as.data.frame(matrix(0, ncol = length(y), nrow = 1))
colnames(m) = y
m[1,] <- x
m <- as.list(m)

getSFS(OutputFile)
?getSFS

#With the HGDP dataset
	data(HGDP.bedassle.data)
str(HGDP.bedassle.data)
str(HGDP.bedassle.data$allele.counts)
#Calculate pairwise Fst between all population pairs
	hgdp.pairwise.Fst <- calculate.all.pairwise.Fst(
		HGDP.bedassle.data$allele.counts,
		HGDP.bedassle.data$sample.sizes
	)	

#Plot pairwise Fst against geographic distance
	plot(HGDP.bedassle.data$GeoDistance,
		hgdp.pairwise.Fst,
		pch=19,
		col=HGDP.bedassle.data$EcoDistance+1,
		ylab="pairwise Fst",
		xlab="geographic distance",
		main="isolation by distance")
			legend(x="bottomright",pch=19,col=c(1,2),
				legend=c("same side of Himalayas",
						"opposite sides of Himalayas"))
```

## Format F0_raw_geno compatible for apparent.R
* column 1 = sample ID
* column 2 = categorical relationship of samples as offspring, parents, etc. (review the apparent package)
* columns 3 until the end = allele IDs with the genotype data as base/base

```{r format F0 Broodstock geno file}

F0_format_geno <- F0_raw_geno %>% 
                      dplyr::mutate(V1 = paste(V1,V2, sep ='_')) %>% 
                      dplyr::select(!c(V2, V3, V4)) %>% 
                      dplyr::rename(Gene.ID = V1)

# sanity check that the IDs file has all we need and everything ran in angsd correctly! 
(ncol(F0_format_geno) -1 ) == nrow(F0_IDs) # msut be TRUE!

# input the IDs as columns 2 through the end of the format_geno
colnames(F0_format_geno)[2:ncol(F0_format_geno)] = F0_IDs$V1

# reformat all genotype calls from basebase to  base/base
for (i in 2:ncol(F0_format_geno)) {
  F0_format_geno[i] = sub("(.{1})(.*)", "\\1/\\2", F0_format_geno[,i])
}


# as matrix 
F0_format_geno.asmatrix <- F0_format_geno
rownames(F0_format_geno.asmatrix)<-F0_format_geno.asmatrix$Gene.ID # make Gene.Id into rownames
F0_format_geno.asmatrix$Gene.ID<-NULL # omit Gene.ID - now you have a matrix with rownames as genes and  cols as sampleiD

# tranpose and format rownames back to the first column as Sample.ID 
F0_transposed_geno <- as.data.frame(t(F0_format_geno.asmatrix)) %>% 
                        tibble::rownames_to_column("Sample.ID")

F0_geno <- F0_transposed_geno

# how many SNPs?
numbSNPs_F0 <- (ncol(F0_geno) - 1)
paste0("Total number of SNPs = ", numbSNPs_F0) # "Total number of SNPs = 37901"```
```


```{r format F1 Juveniles pH 8 geno file}

F1J_pH8_format_geno <- F1J_pH8_raw_geno %>% 
                      dplyr::mutate(V1 = paste(V1,V2, sep ='_')) %>% 
                      dplyr::select(!c(V2, V3, V4)) %>% 
                      dplyr::rename(Gene.ID = V1)

# sanity check that the IDs file has all we need and everything ran in angsd correctly! 
(ncol(F1J_pH8_format_geno) -1 ) == nrow(F1J_pH8_IDs) # must be TRUE!

# input the IDs as columns 2 through the end of the format_geno
colnames(F1J_pH8_format_geno)[2:ncol(F1J_pH8_format_geno)] = F1J_pH8_IDs$V1

# reformat all genotype calls from basebase to  base/base
for (i in 2:ncol(F1J_pH8_format_geno)) {
  F1J_pH8_format_geno[i] = sub("(.{1})(.*)", "\\1/\\2", F1J_pH8_format_geno[,i])
}


# as matrix 
F1J_pH8_format_geno.asmatrix <- F1J_pH8_format_geno
rownames(F1J_pH8_format_geno.asmatrix)<-F1J_pH8_format_geno.asmatrix$Gene.ID # make Gene.Id into rownames
F1J_pH8_format_geno.asmatrix$Gene.ID<-NULL # omit Gene.ID - now you have a matrix with rownames as genes and  cols as sampleiD

# tranpose and format rownames back to the first column as Sample.ID 
F1J_pH8_transposed_geno <- as.data.frame(t(F1J_pH8_format_geno.asmatrix)) %>% 
                        tibble::rownames_to_column("Sample.ID")

F1J_pH8_geno <- F1J_pH8_transposed_geno

# how many SNPs?
numbSNPs_F1JpH8 <- (ncol(F1J_pH8_geno) - 1)
paste0("Total number of SNPs = ", numbSNPs_F1JpH8) # "Total number of SNPs = 52801"
```

```{r format F1 Juveniles pH 75 geno file}

F1J_pH75_format_geno <- F1J_pH75_raw_geno %>% 
                      dplyr::mutate(V1 = paste(V1,V2, sep ='_')) %>% 
                      dplyr::select(!c(V2, V3, V4)) %>% 
                      dplyr::rename(Gene.ID = V1)

# sanity check that the IDs file has all we need and everything ran in angsd correctly! 
(ncol(F1J_pH75_format_geno) -1 ) == nrow(F1J_pH75_IDs) # must be TRUE!

# input the IDs as columns 2 through the end of the format_geno
colnames(F1J_pH75_format_geno)[2:ncol(F1J_pH75_format_geno)] = F1J_pH75_IDs$V1

# reformat all genotype calls from basebase to  base/base
for (i in 2:ncol(F1J_pH75_format_geno)) {
  F1J_pH75_format_geno[i] = sub("(.{1})(.*)", "\\1/\\2", F1J_pH75_format_geno[,i])
}

# as matrix 
F1J_pH75_format_geno.asmatrix <- F1J_pH75_format_geno
rownames(F1J_pH75_format_geno.asmatrix)<-F1J_pH75_format_geno.asmatrix$Gene.ID # make Gene.Id into rownames
F1J_pH75_format_geno.asmatrix$Gene.ID<-NULL # omit Gene.ID - now you have a matrix with rownames as genes and  cols as sampleiD

# tranpose and format rownames back to the first column as Sample.ID 
F1J_pH75_transposed_geno <- as.data.frame(t(F1J_pH75_format_geno.asmatrix)) %>% 
                        tibble::rownames_to_column("Sample.ID")

F1J_pH75_geno <- F1J_pH75_transposed_geno

# how many SNPs?
numbSNPs_F1JpH75 <- (ncol(F1J_pH75_geno) - 1)
paste0("Total number of SNPs = ", numbSNPs_F1JpH75) # "Total number of SNPs = 22766"



#plot the results
nnorm <- function(x) x/sum(x)
getSFS<-function(x) table(factor(rowSums(read.table(x)[,-c(1:4)]),levels=1:20))
YRI=getSFS('HPC_analysis/output/lcWGS/angsd/F1_Juveniles_75/F1Juveniles_pH75_doMaf1_minMaf0.05_majorminor4_minind38_minD5x190_maxD20x760minDind5_maxDind150_minq30_minmapQ.geno.gz')

getSFS(OutputFile)


table(factor(rowSums(OutputFile[,-c(1:4)]),levels=1:20))
read.table('HPC_analysis/output/lcWGS/angsd/F1_Juveniles_75/F1Juveniles_pH75_doMaf1_minMaf0.05_majorminor4_minind38_minD5x190_maxD20x760minDind5_maxDind150_minq30_minmapQ.geno.gz')[,-c(1:4)]

```

## Investigate overlap of SNPs between F0 Broodstock and F1 Juveniles (for fun!)

### use the following files
* F0_format_geno
* F1J_pH8_format_geno
* F1J_pH75_format_geno

* numbSNPs_F0
* numbSNPs_F1JpH8
* numbSNPs_F1JpH75

Obj: determine the number of SNPs that overlap between the juveniles nad their parents 
and what percent of resulting SNPs in offspring derive from parentages, use inner.join
- the reuslting dataframe binds by common contig # and position of the SNP (Gene.ID column)

```{r}
library(plyr)

# :::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::
### F0 and F1 overlap! ::::::::::::::::::::::::::::::::::::::::
# F1 Juveniles pH 8 and F0 Broodstock 
# use inner_join by Gene.ID 
Join_F0_F1JpH8 <- inner_join(F0_format_geno, F1J_pH8_format_geno)
# salitty check for the number of SNPs shared - must be TRUE!
nrow(Join_F0_F1JpH8) == length(intersect(F0_format_geno$Gene.ID, F1J_pH8_format_geno$Gene.ID)) # == TRUE
paste0("Total number of shared SNPs with F1 Juv pH8 = ", nrow(Join_F0_F1JpH8)) # "Total number of SNPs = 17129"
paste0("Percent F1 Juv shared with parents = ", (nrow(Join_F0_F1JpH8)/nrow(F1J_pH8_format_geno))*100 ) # 32.4 % shared!


# F1 Juveniles pH 75 and F0 Broodstock 
# use inner_join by Gene.ID 
Join_F0_F1JpH75 <- inner_join(F0_format_geno, F1J_pH75_format_geno)
# salitty check for the number of SNPs shared - must be TRUE!
nrow(Join_F0_F1JpH75) == length(intersect(F0_format_geno$Gene.ID, F1J_pH75_format_geno$Gene.ID)) # == TRUE
paste0("Total number of shared SNPs with F1 Juv pH8 = ", nrow(Join_F0_F1JpH75)) # "Total number of SNPs = 8123"
paste0("Percent F1 Juv shared with parents = ", (nrow(Join_F0_F1JpH75)/nrow(F1J_pH75_format_geno))*100 ) # 35.7 % shared!


# :::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::
### F1 Treatment overlap ::::::::::::::::::::::::::::::::::::::
# use inner_join by Gene.ID 
Join_F1pH8_F1JpH75 <- inner_join(F1J_pH8_format_geno, F1J_pH75_format_geno)
# salitty check for the number of SNPs shared - must be TRUE!
nrow(Join_F1pH8_F1JpH75) == length(intersect(F1J_pH8_format_geno$Gene.ID, F1J_pH75_format_geno$Gene.ID)) # == TRUE
paste0("Total number of shared SNPs between F1 pH8 and pH75 = ", nrow(Join_F1pH8_F1JpH75)) # "Total number of SNPs = 12988"
paste0("Percent F1 Juv shared between treatments = ", (nrow(Join_F1pH8_F1JpH75)/nrow(F1J_pH75_format_geno))*100 ) # 57.0 % shared!
```



### Fst

```{r}
data(HGDP.bedassle.data)

```
## Parentage datasets 

* format as matric to prepare the merge, include the key (loaded in the start!)
* merge and transpose 
* add type column following apprent criteria 

```{r}

# matrisx + tranpose and format rownames back to the first column as Sample.ID 
# F1Juv_pH75_parentage           <- merge(F0_format_geno, F1J_pH75_format_geno, by="Gene.ID", all = TRUE)
F1Juv_pH75_parentage <- Join_F0_F1JpH75
rownames(F1Juv_pH75_parentage) <- F1Juv_pH75_parentage$Gene.ID # make Gene.Id into rownames
F1Juv_pH75_parentage$Gene.ID   <- NULL
F1Juv_pH75_parentage_df        <- as.data.frame(t(F1Juv_pH75_parentage)) %>% 
                                               tibble::rownames_to_column("Sample.ID")



# matrisx + tranpose and format rownames back to the first column as Sample.ID 
# F1Juv_pH8_parentage           <- merge(F0_format_geno, F1J_pH8_format_geno, by="Gene.ID", all = TRUE)
F1Juv_pH8_parentage <- Join_F0_F1JpH8
rownames(F1Juv_pH8_parentage) <- F1Juv_pH8_parentage$Gene.ID # make Gene.Id into rownames
F1Juv_pH8_parentage$Gene.ID   <- NULL

FJ_pH8_offpsring_match <- as.data.frame(colnames(F1J_pH8_format_geno)[2:ncol(F1J_pH8_format_geno)])



rownames(Join_F1pH8_F1JpH75) <- Join_F1pH8_F1JpH75$Gene.ID # make Gene.Id into rownames
Join_F1pH8_F1JpH75$Gene.ID   <- NULL
x <- as.data.frame(t(Join_F1pH8_F1JpH75))[c(10,56),]

Join_F1pH8_F1JpH75$
sample(nrow(HGDP.bedassle.data$allele.counts),1)

as.data.frame(HGDP.bedassle.data)
FJ_pH8_offpsring_match <- as.data.frame(colnames(F1J_pH8_format_geno)[2:ncol(F1J_pH8_format_geno)])


```


```{r Merge formatted apparent files}


F1Juv_pH8_parentage_df        <- merge(key_F1J_pH8,
                                       (as.data.frame(t(F1Juv_pH8_parentage)) %>% 
                                               tibble::rownames_to_column("Sample.ID"))) %>% 
                                  dplyr::rename(genos = Sample.ID) %>% 
                                  dplyr::filter(!key %in% NA) 
F1Juv_pH8_parentage_df[is.na(F1Juv_pH8_parentage_df)] = "-/-" # this takes a while - a mandatory string for NAs





F1Juv_pH75_parentage_df        <- merge(key_F1J_pH75,
                                       (as.data.frame(t(F1Juv_pH75_parentage)) %>% 
                                               tibble::rownames_to_column("Sample.ID"))) %>% 
                                  dplyr::rename(genos = Sample.ID) %>% 
                                  dplyr::filter(!key %in% NA) 
F1Juv_pH75_parentage_df[is.na(F1Juv_pH75_parentage_df)] = "-/-" # this takes a while - a mandatory string for NAs

```


```{r run apparent}
apparentOUT <- apparent(F1Juv_pH8_parentage_df[,c(1:ncol(F1Juv_pH8_parentage_df))], 
                        MaxIdent=0.10, 
                        alpha=0.01, 
                        nloci=100, 
                        self=TRUE, 
                        plot=FALSE, 
                        Dyad=FALSE)



  GK <- cbind(as.data.frame(F1Juv_pH8_parentage_df[,1]),as.data.frame(F1Juv_pH8_parentage_df[,2]))
  colnames(GK) <- c("genos","key")
  Data <- t(as.data.frame(F1Juv_pH8_parentage_df[,3:ncol(F1Juv_pH8_parentage_df)]))
  ConvertedData <- data.frame (matrix (ncol = ncol(Data), nrow = 0))
  
    for (i in 1:4000) {    
    alleles <- setdiff(strsplit(paste(Data[i,],collapse=""),"")[[1]],c("/","-"))
    pri <- paste(alleles[1],"/",alleles[1],sep="")
    het1 <- paste(alleles[1],"/",alleles[2],sep="")
    het2 <- paste(alleles[2],"/",alleles[1],sep="")
    sec <- paste(alleles[2],"/",alleles[2],sep="")
    
    Data[i,][Data[i,] == pri] <- 0
    Data[i,][Data[i,] == het1 | Data[i,] == het2] <- 0.5
    Data[i,][Data[i,] == sec] <- 1
    Data[i,][Data[i,] == "-/-"] <- NA
    
    line <- as.numeric(as.vector(Data[i,]))
    ConvertedData <- rbind(ConvertedData,line)
  }
  colnames(ConvertedData) <- as.vector(t(GK$genos))
  
  
  Mothers <- vector(mode="numeric",length=0)
  MothersNames <- list()
  Fathers <- vector(mode="numeric",length=0)
  FathersNames <- list()
  Offs <- vector(mode="numeric",length=0)
  OffsNames <- list()
  
    for (i in 1:ncol(ConvertedData)){
    if (GK$key[i] == "Mo") {
      Mothers[i] <- as.data.frame(ConvertedData[,i])
      MothersNames <- append(MothersNames,as.name(as.matrix(GK[i,1])))
      next    
    } else if (GK$key[i] == "Fa") {
      Fathers[i] <- as.data.frame(ConvertedData[,i])
      FathersNames <- append(FathersNames,as.name(as.matrix(GK[i,1])))
      next  
    } else if (GK$key[i] == "Off") {
      Offs[i] <- as.data.frame(ConvertedData[,i])
      OffsNames <- append(OffsNames,as.name(as.matrix(GK[i,1])))
      next
    } else if (GK$key[i] == "Pa") {
      Mothers[i] <- as.data.frame(ConvertedData[,i])
      MothersNames <- append(MothersNames,as.name(as.matrix(GK[i,1])))
      Fathers[i] <- as.data.frame(ConvertedData[,i])
      FathersNames <- append(FathersNames,as.name(as.matrix(GK[i,1])))
      next
    } else if (GK$key[i] == "All") {
      Mothers[i] <- as.data.frame(ConvertedData[,i])
      MothersNames <- append(MothersNames,as.name(as.matrix(GK[i,1])))
      Fathers[i] <- as.data.frame(ConvertedData[,i])
      FathersNames <- append(FathersNames,as.name(as.matrix(GK[i,1])))
      Offs[i] <- as.data.frame(ConvertedData[,i])
      OffsNames <- append(OffsNames,as.name(as.matrix(GK[i,1])))
      next
    } else {
      stop("Please, check the format of the key column (Column 2) in your input file. Acceptable keys for each genotype are Mo, Fa, Off, Pa and All.")
    }
  }
View(F1Juv_pH75_parentage_df)
  
apparentOUT <- apparent(F1Juv_pH75_parentage_df[1:2000], MaxIdent=0.10, alpha=0.05, nloci=100, self=TRUE, plot=TRUE, Dyad=FALSE)

apparentOUT <- apparent(
                        (F1Juv_pH75_parentage_df[1:2000] %>% 
                           dplyr::mutate(key = 'All') %>% 
                           dplyr::filter(!genos %in% c('F1-J256-pH75', 'F1-J354-pH75'))), 
                        MaxIdent=0.10, alpha=0.05, nloci=100, self=TRUE, plot=TRUE, Dyad=FALSE)

apparentOUT$Triad_all
apparentOUT
apparentOUT$Triad_sig
apparentOUT$Triad_summary_pop
apparentOUT$Triad_summary_geno
unique(apparentOUT$Triad_all$Offspring)
unique(apparentOUT$Triad_sig$Offspring)

apparent_editted <- apparentOUT$Triad_all %>% 
  dplyr::filter(!Cross.Type %in% 'self') %>% 
  dplyr::filter(!grepl("F0-B",Offspring),
                grepl("F0-B",Parent1),
                grepl("F0-B",Parent2))

x_final <- data.frame()
  
for (i in 1:nrow(loop_offspring)) {
  
  x <- apparent_editted %>% 
    filter(Offspring %in% loop_offspring[i,]) %>% 
    arrange((GD)) %>% 
    slice(1) 
  
  x_final <- rbind(x, x_final)

}

View(x_final)
```



```{r apparent practice data}
InputFile <- read.table(file="RAnalysis/Scripts/Popgen/apparent_TestData.txt",sep="\t",h=F)


apparentOUT <- apparent(InputFile, MaxIdent=0.10, alpha=0.05, nloci=100, self=TRUE, plot=TRUE, Dyad=FALSE)

apparentOUT
```